ciphersuite/registry.go

```diff
@@ -0,0 +1,160 @@
+package ciphersuite
+
+import (
+	"io"
+
+	"golang.org/x/xerrors"
+)
+
+// Registry stores the cipher suites by name and provides the functions
+// to unpack elements and use the cipher suite primitives.
+type Registry struct {
+	ciphers map[Name]CipherSuite
+}
+
+// NewRegistry creates a new empty registry.
+func NewRegistry() *Registry {
+	return &Registry{
+		ciphers: make(map[Name]CipherSuite),
+	}
+}
+
+// RegisterCipherSuite stores the cipher if it does not exist. It returns the
+// the suite stored for this name if it already exists, or it returns the
+// provided suite.
+func (cr *Registry) RegisterCipherSuite(suite CipherSuite) CipherSuite {
+	name := suite.Name()
+	if suite := cr.ciphers[name]; suite != nil {
+		// Cipher suite already registered so we return it so it can be reused.
+		return suite
+	}
+
+	cr.ciphers[name] = suite
+
+	return suite
+}
+
+func (cr *Registry) get(name Name) (CipherSuite, error) {
+	c, _ := cr.ciphers[name]
+	if c == nil {
+		return nil, xerrors.New("cipher not found")
+	}
+	return c, nil
+}
+
+// UnpackPublicKey takes generic cipher data and tries to convert it
+// into a public key of the associated implementation. The cipher suite
+// must be registered beforehand.
+func (cr *Registry) UnpackPublicKey(raw *RawPublicKey) (PublicKey, error) {
+	c, err := cr.get(raw.Name())
+	if err != nil {
+		return nil, xerrors.Errorf("cipher suite: %v", err)
+	}
+
+	pk, err := c.PublicKey(raw)
```

Comment:
Can't we pass only `raw.Data`?  At this stage we are sure to have the correct cipher suite, so the check `raw.Name() != s.Name()` inside the `PublicKey()` function looks superfluous. Or I am missing a case?